csrc/multidevice/collective.cpp

```diff
@@ -0,0 +1,193 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#ifdef USE_DISTRIBUTED
+
+#include <multidevice/collective.h>
+
+namespace nvfuser {
+
+static inline std::vector<std::vector<at::Tensor>> setBufList(
+    std::vector<at::Tensor> bufs) {
+  return {std::move(bufs)};
+}
+
+static inline void assertBufferCount(
+    const std::vector<at::Tensor>& bufs,
+    size_t count) {
+  TORCH_INTERNAL_ASSERT(
+      bufs.size() == count,
+      "there must be ",
+      count,
+      " buffer(s), but ",
+      bufs.size(),
+      " were given");
+}
+
+static inline void assertBuffersHaveSameSize(
+    const std::vector<at::Tensor>& bufs1,
+    const std::vector<at::Tensor>& bufs2) {
+  if (bufs1.empty() && bufs2.empty()) {
+    return;
+  }
+  auto sizes = (bufs1.empty() ? bufs2 : bufs1).at(0).sizes();
+  for (auto& bufs : {bufs1, bufs2}) {
+    for (auto& buf : bufs) {
+      TORCH_INTERNAL_ASSERT(
+          buf.sizes() == sizes, "all buffers must have the same size");
+    }
+  }
+}
+
+Collective::Collective(CollectiveParams params, std::string name, bool has_root)
+    : params_(std::move(params)),
+      collective_type_(std::move(name)),
+      has_root_(has_root) {
+  assertBuffersHaveSameSize(params_.src_bufs, params_.dst_bufs);
+  TORCH_INTERNAL_ASSERT(
+      std::unique(params_.team.begin(), params_.team.end()) ==
+          params_.team.end(),
+      "the collective must not involve the same device more than once");
+  TORCH_INTERNAL_ASSERT(
+      params_.team.size() > 1, "the team size must be greater than 1");
+  if (has_root_) {
+    auto it = std::find(params_.team.begin(), params_.team.end(), params_.root);
+    TORCH_INTERNAL_ASSERT(
+        it != params_.team.end(),
+        "root (device ",
+        params_.root,
+        ") must be present in the collective's team");
+    root_rank_ = std::distance(params_.team.begin(), it);
```

Comment:
Can you please add some comment why the offset in the team vector is a rank?

Reply:
done, lmk if it's still unclear

Reply:
Well, still not really clear. Is this `root_rank_` a rank or a device? I thought `params_.root` is a device index and `params_.team` is a set of device indices, so `it` here refers to the position of the root device index in the team. Does that mean `root_rank_` is indeed a rank?

Reply:
the variable is now named `root_relative_index_`

Reply:
> Does that mean `root_rank_` is indeed a rank?

what is your definition of rank ?

Reply:
> > Does that mean `root_rank_` is indeed a rank?
> 
> what is your definition of rank ?

That's the question to you.

Reply:
It is just a question of semantics and definition. "rank" is the word used in MPI standard that refers to the unique ID from 0 to <world_size> characterizing a process and, by extension, the process itself. Here we adopt the point of view to not deal with rank directly but with device indices. So even if the notions are very close identical I don't think we should consider this as a rank strictly speaking